apps/sms/test/marionette/lib/messages.js

```diff
@@ -256,6 +256,34 @@
           client.helper.waitForElement(SELECTORS.main);
         },
 
+        launchFromNotification: function(notification) {
+          var chromeClient = client.scope({ context: 'chrome' });
+          chromeClient.executeScript(function(origin, notification) {
+            /* global Components, Services */
+            var managerClass = Components.classes[
+              '@mozilla.org/system-message-internal;1'
+            ];
+
+            var pageURI = Services.io.newURI(
+              origin + '/index.html#notification', null, null
+            );
+
+            var manifestURI = Services.io.newURI(
+              origin + '/manifest.webapp', null, null
+            );
+
+            var systemMessageManager = managerClass.getService(
+              Components.interfaces.nsISystemMessagesInternal
+            );
+
+            systemMessageManager.sendMessage(
+              'notification', notification, pageURI, manifestURI
+            );
+          }, [ORIGIN_URL, notification]);
+
+          this.switchTo();
+        },
+
```

Comment:
so we had a very big discussion here about this, with @lissyx and @JohanLorenzo.

Alexandre and I think it's better to test end to end scenario of:
* receive a message
* 2 scenarios: one with killing the application, one with keeping it
* tap the notification
* proceed from there

To "receive a message" we could use your sendMessage trick. I tried to find out where this was called in gecko, because ideally I'd like to avoid hardcoding the page URI, but I couldn't really find the right place. Maybe @bevis-tseng could help ?

To "tap the notification" there is some code in the system app, not really elegant but at least it's there. Likely we should reuse https://github.com/mozilla-b2g/gaia/blob/master/apps/system/test/marionette/lib/notification.js, especially the object `NotificationList` so that it's easy to tap on the right notification.